scheduler/pkg/envoy/xdscache/seldoncache.go

```diff
@@ -91,109 +197,168 @@ func (xds *SeldonXDSCache) SetupTLS() error {
+	return []types.Resource{defaultRoutes, mirrorRoutes}
+}
+
+func (xds *SeldonXDSCache) UpdateRoutes(nodeId string) error {
+	logger := xds.logger.WithField("func", "UpdateRoutes")
+	// Create the snapshot that we'll serve to Envoy
+	snapshot, err := cache.NewSnapshot(
+		xds.newSnapshotVersion(), // version
+		map[resource.Type][]types.Resource{
+			resource.RouteType: xds.RouteResources(), // routes
+		})
+	if err != nil {
+		logger.Errorf("could not create snapshot %+v", snapshot)
+		return err
+	}
+
+	logger.Debugf("will serve snapshot %+v", snapshot)
+
+	// Add the snapshot to the cache
+	if err := xds.snapshotCache.SetSnapshot(context.Background(), nodeId, snapshot); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (xds *SeldonXDSCache) AddClusters() error {
 	var clientSecret *Secret
-	if xds.TLSActive {
-		if secret, ok := xds.Secrets[EnvoyUpstreamClientCertName]; ok {
+	if xds.tlsActive {
+		if secret, ok := xds.secrets[envoyUpstreamClientCertName]; ok {
 			clientSecret = &secret
 		}
 	}
 
-	r = append(r, xds.permanentClusters...)
+	resources := make(map[string]types.Resource)
+	for clusterName := range xds.clustersToAdd {
+		cluster, ok := xds.Clusters[clusterName]
+		if ok {
+			resource := makeCluster(cluster.Name, cluster.Endpoints, cluster.Grpc, clientSecret)
+			resources[cluster.Name] = resource
 
-	for _, c := range xds.Clusters {
-		endpoints := make([]Endpoint, 0, len(c.Endpoints))
-		for _, value := range c.Endpoints { // Likely to be small (<100?) as is number of model replicas
-			endpoints = append(endpoints, value)
 		}
-		r = append(r, MakeCluster(c.Name, endpoints, c.Grpc, clientSecret))
+		delete(xds.clustersToAdd, clusterName)
 	}
 
-	return r
-}
-
-func (xds *SeldonXDSCache) RouteContents() []types.Resource {
-	defaultRoutes, mirrorRoutes := MakeRoutes(xds.Routes, xds.Pipelines)
-	return []types.Resource{defaultRoutes, mirrorRoutes}
+	return xds.cds.UpdateResources(resources, nil)
 }
 
-func (xds *SeldonXDSCache) ListenerContents() []types.Resource {
-	return xds.permanentListeners
-}
-
-func (xds *SeldonXDSCache) SecretContents() []types.Resource {
-	logger := xds.logger.WithField("func", "SecretContents")
-	var r []types.Resource
-
-	for _, s := range xds.Secrets {
-		secrets := MakeSecretResource(s.Name, s.ValidationSecretName, s.Certificate)
-		logger.Infof("Adding secrets for %s(%s) of length %d", s.Name, s.ValidationSecretName, len(secrets))
-		for _, secret := range secrets {
-			r = append(r, secret)
+func (xds *SeldonXDSCache) RemoveClusters() error {
+	clustersToRemove := make([]string, 0)
+	for clusterName := range xds.clustersToRemove {
+		if xds.shouldRemoveCluster(clusterName) {
+			clustersToRemove = append(clustersToRemove, clusterName)
 		}
+		delete(xds.clustersToRemove, clusterName)
 	}
 
-	return r
+	return xds.cds.UpdateResources(nil, clustersToRemove)
+}
+
+func (xds *SeldonXDSCache) shouldRemoveCluster(name string) bool {
```

Comment:
nit: maybe describe a case where a cluster is deleted but it does still have routes? Is this now possible given that clusters have their names with a model+version? Probably safer to keep anyway in case behaviour changes in the future.

Reply:
Because updates are batched, the state might change between the update occurring and when it's actually sent to envoy.